### ISSUE TITLE
False positible ANN001 Missing type annotation for function argument `self`

### ISSUE THREAD

**Issue by ssbarnea** [astral-sh/ruff#14295]
Current version of ruff (0.7.3) does provide a false-positive message about types for `self`.

Config reproduce
```toml
[tool.ruff.lint]
select = ["ALL"]
```

While [ANN101 was removed more than a year ago](https://github.com/astral-sh/ruff/issues/4396) and its issue closed as done, we missed to spot that `ANN001` will trigger the same problem.

That is forcing us to disable entire `ANN001` rule until this is sorted as adding type to self does create a huge amount of boilerplate that is only making function signatures harder to read.

**Comment by MichaReiser** [comment]
@AlexWaygood I'm a bit at a loss here. Would you mind taking a look?

**Comment by dhruvmanila** [comment]
Can you provide a code to reproduce this? I believe it shouldn't flag the `self` argument: https://play.ruff.rs/df6e79ac-202b-454e-b85e-120619b96645

Looking at the implementation, it does skip `self` or `cls` for non-static methods:

https://github.com/astral-sh/ruff/blob/be69f61b3ec42f5cda2490ff64c1efb9814f98f0/crates/ruff_linter/src/rules/flake8_annotations/rules/definition.rs#L605-L612

**Comment by ssbarnea** [comment]
Oops, sorry my mistake. It proved to be a nested function with the very unfortunate (incorrect) use of self.
